processor/batchprocessor/splitmetrics.go

```diff
@@ -0,0 +1,65 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package batchprocessor
+
+import (
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+// splitMetrics removes metrics from the input data and returns a new data of the specified size.
+func splitMetrics(size int, toSplit pdata.Metrics) pdata.Metrics {
+	if toSplit.MetricCount() <= size {
+		return toSplit
+	}
+	copiedMetrics := 0
+	result := pdata.NewMetrics()
+	rms := toSplit.ResourceMetrics()
+	for i := rms.Len() - 1; i >= 0; i-- {
+		rm := rms.At(i)
+		destRs := pdata.NewResourceMetrics()
+		rm.Resource().CopyTo(destRs.Resource())
+		result.ResourceMetrics().Append(destRs)
```

Comment:
This can be improved to avoid unnecessary allocations.

```
resultLen := result.ResourceMetrics().Len()
result.ResourceMetrics().Resize(rms.Len())
i := rms.Len() - 1
for ; i >= 0 && copiedMetrics < size; i-- {
  destRs := result.ResourceMetrics().At(rms.Len() - 1 - i)
  ...
}
// Please confirm if it is i+1 or i :) 
result.ResourceMetrics().Resize(i+1)
return result
```

Reply:
This doesn't work because we can't know how many metrics will be in each ResourceMetrics before looping through them. The size is the number of total metrics not the number of ResourceMetrics. We could precompute the needed sizes by doing one loop to calculate sizes and do the resizes and another loop to do the actual copies?

Reply:
I am confused a bit, the size of the Resources/Libraries/Metrics is at most the number of current Resources/Libraries/Metrics that you are cloning, so that is a good estimate. We can leave it as is and leave a TODO.

Reply:
Ah. I thought you were suggesting to resize to the "correct" size at the start. Seems you were suggesting to resize to the max size at the start, then resize back down at the end?

Reply:
Yes

Reply:
Done in 711767a